src/devices/markisol.c

```diff
@@ -0,0 +1,136 @@
+/** @file
+    Markisol (a.k.a E-Motion, BOFU, BF-30x, BF-415) curtains remote.
+
+    Copyright (C) 2021 Dan Stahlke <dan@stahlke.org>
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+*/
+
+/**
+Markisol curtains remote.
+
+Protocol description:
+Each frame starts with:
+    hi 4886us
+    lo 2470us
+    hi 1647us
+    lo 315us
+Then follow 40 bits:
+    zero: hi 670us, lo 320us
+    one : hi 348us, lo 642us
+
+This is OOK_PULSE_PWM encoding.  The frame is erroneously interpred as a bit (so bitbuffer_t reports
+41 bits rather than 40).  We discard this bit during recording.  The last frame erroneosly picks up
+an extra bit at the end; we ignore this as well.
+
+Packet interpretation:
+    16 bits - unique ID of remote
+    16 bits - channel, zone, and control
+    8  bits - checksum (all bytes, including this one, sum to 1)
+
+The second pack of 16 bits is interwoven:
+    buf[2] & 0x0f - channel, in the range 1-15
+    buf[2] & 0x20 - bit 0 of zone
+    buf[2] & 0xd0 - bits 0,2,3 of control
+    buf[3] & 0x10 - bit 1 of control
+    buf[3] & 0x80 - bit 1 of zone
+    buf[3] & 0x6f - unknown; for my remotes (buf[3] & 0x6f) == 0x01 always
+*/
+
+#include "decoder.h"
+
+static int markisol_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    uint8_t buf[5];
+    uint8_t cksum = 0;
+    int got_proper_row_length = 0;
+    for (int i = 0; i < bitbuffer->num_rows; i++) {
+        if (decoder->verbose > 0)
+            fprintf(stderr, "%s: bits_per_row[%d] = %d\n", __func__, i, bitbuffer->bits_per_row[i]);
+        if (bitbuffer->bits_per_row[i] == 41 || bitbuffer->bits_per_row[i] == 42) {
+            uint8_t *b = bitbuffer->bb[i];
+            for (int j = 0; j < 5; ++j) {
+                buf[j] = ~reverse8(((b[j] << 8) + b[j+1]) >> 7);
```

Comment:
This should have been `<< 1` not `<< 8`, right?

Reply:
I was joining consecutive bytes into a 16-bit value, then shifting the whole thing.  But <<1 is probably clearer.  I rewrote it and added a comment.

Reply:
Oh, I messed up counting the parens, you are right, either way works.